### ISSUE TITLE
Troubleshooting iOS background audio playback

### ISSUE THREAD

**Issue by drew-512** [RenderHeads/UnityPlugin-AVProVideo#1243]
- Unity 2021.3.7
- AVPro Core 2.5.7 (1282)
- iOS 15.4

Hi AVPro friends, 

We got AVPro with the intent to use it as an audio playback engine for a Unity based audio player.  

Currently, AVPro seems to auto-pause when the iOS Unity app goes to background.  I've double-checked the Unity background settings, and get Xcode runtime output when I move the app to background:

```
-> applicationWillResignActive()
-> applicationDidEnterBackground()
2022-08-02 16:23:03.538976-0500 PlanetPlayer[22052:11704175] [Player] player rate changed to 0.000000
-> applicationWillEnterForeground()
-> applicationDidBecomeActive()
2022-08-02 16:23:13.056482-0500 PlanetPlayer[22052:11704175] [Player] player rate changed to 1.000000
```

...which seems to suggest that AVPro is pausing the audio, whereas we want to have full control over that (and not pause when going to background).   My attempts to find where this auto-pausing is controlled have been unsuccessful, so they are presumably in the AVPro binaries and outside our control.

Also, best practices and recommendations to produce a stable, happy audio player and kindly invited.  💙

Thanks and keep up the great work!  

Drew

**Comment by MorrisRH** [comment]
Hey @drew-512, pausing playback is standard iOS behaviour when you move an app to the background. In order to have playback continue you will need to do the following:
1. Add the background audio capability to your application in the Xcode project, details can be found [here](https://developer.apple.com/documentation/avfoundation/media_playback/creating_a_basic_video_player_ios_and_tvos/enabling_background_audio?language=objc)
2. Change the AVAudioSession category to "playback", details [here](https://developer.apple.com/documentation/avfaudio/avaudiosession?language=objc)
3. Make sure the MediaPlayer component has "Audio Mode" set to "System Direct" in the iOS platform settings

The only awkward part is point 2 as it will require native code. You can do this directly in the generated Xcode project by editing the method ```- (void)startUnity:(UIApplication*)application``` in UnityAppController.mm (approx. line 117) and replacing:
```objc
[audioSession setCategory: AVAudioSessionCategoryAmbient error: nil];
```
with:
```objc
[audioSession setCategory: AVAudioSessionCategoryPlayback error: nil];
```

**Comment by drew-512** [comment]
Lovely, thank you sir.

Any other best practices that come to mind would be appreciated.  We've been authoring realtime AV software on native iOS, OS X, and Windows for 10 years now, so doing things anew in Unity is a little scary -- glad to have AVPro at our backs!

Drew

**Comment by drew-512** [comment]
Hi @MorrisRH,

I made these changes and a `UnityAppController.mm` patch for `AVAudioSessionCategoryPlayback` and had success.   

However, since this bypasses Unity's `AudioSource` (for System Direct), I'm not seeing a way to get FFT data out, ugh.   This is a major issue since having fancy audio visualizers is a core value-add of our product.

Is there any way to get PCM audio out of AVPro?   We have our own FFT analyzers and filters, but for this approach, I'd need a live PCM tap so I can do the processing and then hand that off to our visualization subsystem.  Perhaps I could intercept where AVPro pass audio to iOS and use that?  That of course resides within AVPro's iOS glue bundle, and so I have no way to access that.  Are there any solutions I'm missing?

We've made music visuals for almost two decades, so it'd be quite a setback if we can't navigate this and would have to reevaluate our audio engine options.   Fun fact is I made some of the original Winamp visualizers that made their way into iTunes back in the day, so I'd love to figure this out with you and happy to take this to email.

Thanks,
Drew

**Comment by MorrisRH** [comment]
Hey @drew-512
The Unity audio output mode will deliver what you need to be able to run your FFT analysers, however you will lose support for background audio playback. There's currently no way to switch between the two modes at runtime although this may be possible but will require some investigation from our end.

**Comment by drew-512** [comment]
Ok, I figured as much.

Well, I've decades of work on the native iOS and desktop realtime audio and graphics (see previous links), so I'm open minded on how I can contribute or work with you.

Drew

**Comment by drew-512** [comment]
Hi @MorrisRH,

Sorry if this is a dumb question but what is AVPro's `GrabAudio(float[] buffer, int sampleCount, int channelCount)`?   Couldn't I use that as an audio tap to do necessary audio visulization?

Thanks,
Drew

**Comment by MorrisRH** [comment]
We have a mini spectrum visualiser built into our media player demo scene. If you look at MediaPlayerUI.cs and search for the method named "UpdateAudioSpectrum" you'll see how we're doing it.

**Comment by drew-512** [comment]
Hi, yes I saw that.  However, that seems to just pull FFT data from Unity's `AudioSource` each frame and pump that into AVPro's little shader.   Of course that updating stops once AVPro is set to use the audio engine.

`GrabAudio()` on the other hand appears to be exactly what we're looking for (live PCM data that can use in our visual engine).

I suspect `GrabAudio()` was once intended to be used within `UpdateAudioSpectrum()` when not using `AudioSource` but for whatever reason that code is not present.  It would be lovely that `GrabAudio()` does the trick but I didn't want to hang our hat on that it's dead or on it way out.

**Comment by MorrisRH** [comment]
I think reading data from the AudioSource using GetOutputData() is your best bet then.
GrabAudio is our route for pulling audio data back from the native side of the plugin and feeding it into Unity. Were you to use it you would break how our audio playback components work (of course you could reimplement these yourselves).

**Comment by MorrisRH** [comment]
I've had a play with things and I think I can give you the best of both worlds; being able to get you the audio data via GrabAudio and allowing background playback.

If you can send an email to unitysupport@renderheads.com with the following subject "#1243 Troubleshooting iOS background audio playback", I'll send you a build with the necessary changes.

**Comment by Chris-RH** [comment]
Thanks @drew-512, email received :)

**Comment by MorrisRH** [comment]
Closing this issue as now resolved with version 2.6.0.
